smartmoneyconcepts/new_smc.py

```diff
@@ -0,0 +1,1180 @@
+                if _low[i] <= last_swing_low:
+
+                    if trend == -1:
+                        bos[last_swing_low_index] = -1
+                        if use_bos:
+                            use_structure = True
+                    else:
+                        choch[last_swing_low_index] = -1
+                        trend = -1
+                        if use_choch:
+                            use_structure = True
+
+                    bos_level[last_swing_low_index] = last_swing_low
+                    structure_complete_date[last_swing_low_index] = _dates[i]
+
+                    if use_structure:
+                        ob[last_swing_high_index] = 1
+                        ob_confirm_date[last_swing_high_index] = _dates[i]
+
+                        ob_high_indices = np.concatenate(
+                            (
+                                ob_high_indices,
+                                np.array([[last_swing_high_index, last_swing_high]]),
+                            ),
+                            axis=0,
+                        )
+
+                    if last_swing_high < last_ob_high:
+                        last_ob_high = last_swing_high
+
+                    last_swing_low = 0
+                    last_swing_low_index = 0
+
+                # lower order block mitigated
+                if _low[i] <= last_ob_low:
+
+                    idx_to_delete = 0
+                    for row_idx, row in enumerate(ob_low_indices):
+                        ob_low_index, swing_low = row
+                        if last_ob_low == swing_low:
+                            idx_to_delete = row_idx
+                            ob_mitigation_date[int(ob_low_index)] = _dates[i]
+
+                    mask = np.ones(ob_low_indices.shape[0], dtype=bool)
+                    mask[idx_to_delete] = False
+                    ob_low_indices = ob_low_indices[mask]
+
+                    if np.any(ob_low_indices):
+                        last_ob_low = ob_low_indices[:, 1].max()
+                    else:
+                        last_ob_low = very_low_no
+
+                if _high_low[i] == 1:
+                    last_swing_high = _high[i]
+                    last_swing_high_index = i
+
+                if _high_low[i] == -1:
+                    last_swing_low = _low[i]
+                    last_swing_low_index = i
+
+        # replace all the 0s with np.nan
+        bos = np.where(bos != 0, bos, np.nan)
+        choch = np.where(choch != 0, choch, np.nan)
+        bos_level = np.where(bos_level != 0, bos_level, np.nan)
+        ob = np.where(ob != 0, ob, np.nan)
+
+        top = np.where(np.isnan(ob), np.nan, _high)
+        bottom = np.where(np.isnan(ob), np.nan, _low)
+
+        ob = pd.Series(ob, name="OB")
+        top_series = pd.Series(top, name="Top")
+        bottom_series = pd.Series(bottom, name="Bottom")
+        ob_confirm_date = pd.Series(ob_confirm_date, name="ConfirmDate")
```

Comment:
@CosmicTrader I would suggest to keep consistency with the rest of the indicators where they actually refer to the "Index" and not a date/timestamp. In the end, the index can be either a timestam or just a number dpeending on which data you are using, so I woulkd prefer to have it more generic to cover more use cases. We could also add both, but I think it is an overload.